### PR TITLE
ENH: add get_snapshot

### PR DESCRIPTION
**Pull request by klauer**
Add `get_snapshot` tool to get PV data (one or more PVs or globs) at a given point in time.

* Added pre-commit config
* Started `get_snapshot` utility for getting values of multiple PVs at a specific point in time (referred to as save/restore API in the docs)
* Cleaned up some flake8 issues (sorry for the mess)
* To investigate: skipped `search()` to pattern match via globs if no glob characters were included. Ensure this doesn't mess up any current functionality.

WIP until Murali gets around to adding metadata to the snapshot data, which is estimated to be live next month.

### CODE REVIEW

**Comment by klauer** [comment]
Going against what I said in the PR text, I don't think we really _need_ Murali's archiver update to get this in. 

It does work as-is, just does not contain all the metadata that we want in the `get_snapshot` response. I'll ask for a review and we'll go from there.

File: archapp/interactive.py [hunk 1]
```diff
@@ -151,22 +155,58 @@ def get(
...
+
+        Parameters
+        ----------
+        pvname : str, list/tuple of str
         """
         if isinstance(pvname, str):
-            return self.search(pvname, do_print=False)
-        elif isinstance(pvname, (list, tuple)):
+            if "*" in pvname or "?" in pvname:
+                return self.search(pvname, do_print=False)
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply archapp/interactive.py:198]: Yes, this is worth doing and good. One small thing to check is the behavior change when we give an unarchived PV. The previous implementation returned an empty list, this one will return a list of one string.

File: archapp/interactive.py [hunk 2]
```diff
@@ -258,15 +301,9 @@ def convert_date_arg(arg, unit):
     date : datetime
     """
     if isinstance(arg, (list, tuple)):
-        try:
-            return dt.datetime(*arg)
-        except:
-            raise
```
Comment (ZLLentz) [inline archapp/interactive.py:264]: oh dear

**Review by ZLLentz — APPROVED** [review]
Additions and fixes look good, post request matches the archiver appliance docs

File: archapp/interactive.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (klauer) [reply archapp/interactive.py:203]: I walked this back a bit for `.get()`. I'm not 100% sure of what downstream clients expect out of `.get()` without existing PVs.

As of my latest commit, this PR results in the following:
```
In [5]: arch.get("XPP:USR:MMS:01", dt.datetime(2016, 11, 10), dt.datetime(2016, 11, 11))
Out[5]:
array([['2016-11-09T19:27:01.490605056', -4.882809997752702e-05],
       ['2016-11-10T11:27:04.032083712', 5.0],
       ['2016-11-10T11:32:43.301595392', 0.0],
       ...,
       ['2016-11-10T21:48:59.198031104', -9.210639866600001],
       ['2016-11-10T21:48:59.908970752', -9.210639866600001],
       ['2016-11-10T21:49:00.335118336', -9.210639866600001]],
      dtype=object)

In [6]: arch.get("XPP:USR:MMS:ABC", dt.datetime(2016, 11, 10), dt.datetime(2016, 11, 11))
Out[6]: array([], dtype=float64)

In [7]: arch.get([ "A", "B", "C", "AT1K4:L2SI:MMS:01.RBV", "AT1K4:L2SI:MMS:00.RBV", "AT1K4:L2SI:MMS:02.RBV", ], start=datetime.datetime.now(), end=
   ...: datetime.datetime.now(), xarray=True)
Out[7]:
<xarray.Dataset>
Dimensions:                (time: 3, field: 3)
Coordinates:
  * time                   (time) datetime64[ns] 2021-12-01T12:02:39.18141260...
  * field                  (field) <U4 'vals' 'sevr' 'stat'
Data variables:
    AT1K4:L2SI:MMS:01.RBV  (field, time) object 22.70155 nan 22.7016 ... 0 nan 0
    AT1K4:L2SI:MMS:02.RBV  (field, time) object nan 24.0098 nan ... nan 0 nan
Attributes:
    name:     AT1K4:L2SI:MMS:01.RBV
    EGU:      mm
    PREC:     3
```

`get_snapshot` is new API, and a single request is just fine for it; so a flag was added for "should I hit the management port for every PV to see if it's there before asking for data"

[the remark this answers is not in the document]
Comment (ZLLentz) [reply archapp/interactive.py:203]: Seems appropriately cautious and valid :+1:

**Comment by ZLLentz** [comment]
Environment variable additions look good too! I forgot this was so hard-coded

**Pull request merged**